File: Find K Smallest Sum Pairs/Find K Smallest Sum Pairs.cpp
Best Video Solution with Live Coding: https://youtu.be/h_RfV9NYihc
LIKE | SHARE | SUBSCRIBE


klog, k

class Solution {
public:
    vector<vector<int>> kSmallestPair(vector<int>& arr1, vector<int>& arr2, int k) {
        vector<vector<int>> result;
        
        if(arr1.empty() || arr2.empty() || k == 0) return result;

        // Min Heap storing {sum, i, j}
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;

        // Push first elements from each row
        for(int i = 0; i < min((int)arr1.size(), k); i++) {
            pq.push({arr1[i] + arr2[0], i, 0});
        }

        while(!pq.empty() && k > 0) {
            auto top = pq.top();
            pq.pop();

            int i = top[1];
            int j = top[2];

            result.push_back({arr1[i], arr2[j]});
            k--;

            // Move right in arr2
            if(j + 1 < arr2.size()) {
                pq.push({arr1[i] + arr2[j + 1], i, j + 1});
            }
        }

        return result;
    }
};



Find K Smallest Sum Pairs

File: Find K Smallest Sum Pairs/Find K Smallest Sum Pairs.java
Best Video Solution with Live Coding: https://youtu.be/h_RfV9NYihc
LIKE | SHARE | SUBSCRIBE


  klog, k
  class Pair {
    int i, j, sum;
    Pair(int i, int j, int sum) {
        this.i = i;
        this.j = j;
        this.sum = sum;
    }
}


class Solution {
    public ArrayList<ArrayList<Integer>> kSmallestPair(int[] arr1, int[] arr2, int k) {
       ArrayList<ArrayList<Integer>> res = new ArrayList<>();
       
       PriorityQueue<Pair> pq = new PriorityQueue<>((a,b) -> a.sum -b.sum);
       
       for(int i=0; i<Math.min(arr1.length,k); i++){
       pq.add(new Pair(i,0, arr1[i] + arr2[0]));
        }
    
    while(!pq.isEmpty() && k>0){
        Pair p = pq.poll();
        ArrayList<Integer> curr = new ArrayList<>();
        curr.add(arr1[p.i]);
         curr.add(arr2[p.j]);
         res.add(curr);
         
         k--;
         
         // push
         if(p.j + 1 < arr2.length){
              pq.add(new Pair(p.i, p.j + 1, arr1[p.i] + arr2[p.j + 1]));
         }
        
    }
    return res;

    }
}

